File: C++/permuteUnique.cpp
/**
 * @File    : permuteUnique.cpp
 * @Brief   : 给定一个可包含重复数字的序列 nums ，按任意顺序 返回所有不重复的全排列。
 * @Link    : https://leetcode-cn.com/problems/permutations-ii/
 * @Author  : Wei Li
 * @Date    : 2021-07-12
*/

#include <iostream>
#include <vector>
#include <algorithm>

// 方法一：搜索回溯
class Solution
{
    std::vector<int> vis;

public:
    void backtrack(std::vector<int> &nums, std::vector<std::vector<int>> &ans, int idx, std::vector<int> &perm)
    {
        if (idx == nums.size())
        {
            ans.emplace_back(perm);
            return;
        }
        for (int i = 0; i < (int)nums.size(); ++i)
        {
            if (vis[i] || (i > 0 && nums[i] == nums[i - 1] && !vis[i - 1]))
            {
                continue;
            }
            perm.emplace_back(nums[i]);
            vis[i] = 1;
            backtrack(nums, ans, idx + 1, perm);
            vis[i] = 0;
            perm.pop_back();
        }
    }

    std::vector<std::vector<int>> permuteUnique(std::vector<int> &nums)
    {
        std::vector<std::vector<int>> ans;
        std::vector<int> perm;
        vis.resize(nums.size());
        sort(nums.begin(), nums.end());
        backtrack(nums, ans, 0, perm);
        return ans;
    }
};

template <typename T>
void DisplayElement(const T &container)
{
    for (auto element = container.begin(); element != container.end(); ++element)
    {
        for (auto value = element->begin(); value != element->end(); ++value)
        {
            std::cout << *value << " ";
        }
        std::cout << std::endl;
    }
    std::cout << std::endl;
}

// ------------------------------------
int main(int argc, char **argv)
{
    // std::vector<int> nums{1, 1, 2};
    std::vector<int> nums{1, 2, 3};

    Solution solution;
    auto permute_unique_vec = solution.permuteUnique(nums);
    std::cout << "The solutio of this problem is : " << std::endl;
    DisplayElement(permute_unique_vec);

    return 0;
}

File: Python/permuteUnique.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: permuteUnique.py
@Function: 给定一个可包含重复数字的序列 nums ，按任意顺序 返回所有不重复的全排列。
@Link: https://leetcode-cn.com/problems/permutations-ii/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-12
"""


class Solution:
    def permuteUnique(self, nums):
        nums.sort()
        self.res = []
        check = [0 for i in range(len(nums))]

        self.backtrack([], nums, check)
        return self.res

    def backtrack(self, sol, nums, check):
        if len(sol) == len(nums):
            self.res.append(sol)
            return

        for i in range(len(nums)):
            if check[i] == 1:
                continue
            if i > 0 and nums[i] == nums[i-1] and check[i-1] == 0:
                continue
            check[i] = 1
            self.backtrack(sol+[nums[i]], nums, check)
            check[i] = 0


# -------------------------
if __name__ == "__main__":
    # nums = [1, 1, 2]
    nums = [1, 3, 2]

    solution = Solution()
    permute_unique_list = solution.permuteUnique(nums=nums)
    print(f"The solution of this problem is : {permute_unique_list}")
